plugins/mongo_db/mongo_db_writer.cpp

```diff
@@ -0,0 +1,315 @@
+            block_doc << operations << operations_array;
+        }
+
+        static const std::string transactions = "transactions";
+        block_doc << transactions << transactions_array;
+
+        static const std::string blocks = "blocks";
+        if (formatted_blocks.find(blocks) == formatted_blocks.end()) {
+            formatted_blocks[blocks] = std::make_unique<mongocxx::bulk_write>(bulk_opts);
+        }
+
+        mongocxx::model::insert_one insert_msg{block_doc.view()};
+        formatted_blocks[blocks]->append(insert_msg);
+    }
+
+    void mongo_db_writer::write_document(named_document const& named_doc) {
+        if (formatted_blocks.find(named_doc.collection_name) == formatted_blocks.end()) {
+            formatted_blocks[named_doc.collection_name] = std::make_unique<mongocxx::bulk_write>(bulk_opts);
+        }
+
+        //auto view = named_doc.doc.view();
+        //auto itr = view.find("_id");
+        //if (view.end() == itr) {
+        //    mongocxx::model::insert_one msg{std::move(view)};
+        //    formatted_blocks[named_doc.collection_name]->append(msg);
+        //} else {
+            document filter;
+
+            filter << "_id" << bsoncxx::oid(named_doc.keyval); 
+
+            mongocxx::model::update_one msg{filter.view(), 
+                named_doc.doc.view()};
+            msg.upsert(true);
+
+            if (indexes.find(named_doc.collection_name) == indexes.end()) {
+                 for (auto& index_to_create : named_doc.indexes_to_create) {
+                    mongo_database[named_doc.collection_name].create_index(index_to_create.view());
+                 }
+                 indexes[named_doc.collection_name] = "created";
+            }
+
+            formatted_blocks[named_doc.collection_name]->append(msg);
+        //}
+    }
+
+    void mongo_db_writer::remove_document(named_document const& named_doc) {
+        if (formatted_blocks.find(named_doc.collection_name) == formatted_blocks.end()) {
+            formatted_blocks[named_doc.collection_name] = std::make_unique<mongocxx::bulk_write>(bulk_opts);
+        }
+
+        //mongo_database[named_doc.collection_name].update_many(
+        //    document{} << key << bsoncxx::oid(oid_hash) << finalize,
+        //    document{} << "$set" << open_document << "removed" << true << close_document << finalize);
+        
+        document filter;
+        filter << named_doc.key << bsoncxx::oid(named_doc.keyval);
+        auto v1 = filter.view();
+        document newval;
+        newval << "$set" << open_document << "removed" << true << close_document;
+        auto v2 = newval.view();
+        mongocxx::model::update_many msg{v1, v2};
+        formatted_blocks[named_doc.collection_name]->append(msg);
+        
+        // Or, to remove permanently:
+        //mongo_database[named_doc.collection_name].delete_many(
+        //    document{} << "_id" << bsoncxx::oid(oid_hash) << finalize);
```

Comment:
Lets keep these lines. It can help in the future, when we will add configuration for delete/update removed objects.